File: DigitalImageProcessing/cpp/fourier_transform.cpp
/**
 * @File    : fourier_transform.cpp
 * @Brief   : 二维离散傅里叶变换; 快速傅里叶变换; 幅度谱(零谱中心化)和相位谱
 * @Author  : Wei Li
 * @Date    : 2021-10-21
*/


#include <iostream>
#include <vector>
#include <opencv2/core.hpp>
#include <opencv2/imgproc.hpp>
#include <opencv2/highgui.hpp>


// ------------ 快速离散傅里叶变换 ------------
void FFT2Image(cv::InputArray InputImage, cv::OutputArray img_fourier)
{
    cv::Mat image = InputImage.getMat();
    int rows = image.rows;
    int cols = image.cols;

    // 满足快速傅里叶变换的最优 行数 和 列数
    int row_padding = cv::getOptimalDFTSize(rows);
    int col_padding = cv::getOptimalDFTSize(cols);
    // 左侧和下侧 zero-padding
    cv::Mat fourier;
    cv::copyMakeBorder(image, fourier, 0, row_padding - rows, 0, col_padding - cols, cv::BORDER_CONSTANT, cv::Scalar::all(0));

    // 快速傅里叶变换
    cv::dft(fourier, img_fourier, cv::DFT_COMPLEX_OUTPUT);
}


// ------------ 傅里叶变换的两个度量: 幅度谱和相位谱 ------------
void AmplitudeSpectrumFFT(cv::InputArray _srcFFT, cv::OutputArray _dstSpectrum)
{
    // 实部和虚部两个通道
    CV_Assert(_srcFFT.channels() == 2);
    // 分离实部和虚部两个通道
    std::vector<cv::Mat> FFT2channels;
    cv::split(_srcFFT, FFT2channels);
    // compute magnitude spectrum of FFT
    cv::magnitude(FFT2channels[0], FFT2channels[1], _dstSpectrum);
}

// 对于傅里叶谱的灰度级显示，OpenCV 提供了函数 log，
// 该函数可以计算矩阵中每一个值的对数。
// 进行归一化后，为了保存傅里叶谱的灰度级，有时需要将矩阵乘以 255，然后转换为 8 位图。
cv::Mat graySpectrum(cv::Mat spectrum)
{
    cv::Mat dst;
    cv::log(spectrum + 1, dst);
    cv::normalize(dst, dst, 0, 1, cv::NORM_MINMAX);
    // 为了灰度级可视化
    dst.convertTo(dst, CV_8UC1,  255, 0);

    return dst;
}

// OpenCV function : cv::phase(x, y, angle, angleInDegress);
cv::Mat PhaseSpectrum(cv::Mat _srcFFT)
{
    cv::Mat phase_spectrum;
    phase_spectrum.create(_srcFFT.size(), CV_64FC1);

    std::vector<cv::Mat> FFT2channels;
    cv::split(_srcFFT, FFT2channels);

    // 计算相位谱
    for (size_t r = 0; r < phase_spectrum.rows; ++r)
    {
        for (size_t c = 0; c < phase_spectrum.cols; ++c)
        {
            double real_part = FFT2channels[0].at<double>(r, c);
            double imaginary_part = FFT2channels[1].at<double>(r, c);
            // atan2 返回值范围 [0, 180], [-180, 0]
            phase_spectrum.at<double>(r, c) = std::atan2(imaginary_part, real_part);
        }
    }

    return phase_spectrum;
}


// -------------------------------
int main(int argc, char** argv)
{
    if (argc > 1)
    {
        cv::Mat image = cv::imread(argv[1], 0);
        if (!image.data)
        {
            std::cout << "Error: reading image unsuccesfully." << std::endl;
            return -1;
        }
        cv::imshow("OriginImage", image);

        // ------------ 离散傅里叶变换 ------------
        // 输入类型转换 CV_8U ----> CV_32F or CV_64F
        cv::Mat image_float;
        image.convertTo(image_float, CV_64F);

        // Fourier Transform
        cv::Mat img_fourier;
        cv::dft(image_float, img_fourier, cv::DFT_COMPLEX_OUTPUT);
        // Invert Fourier Transform, only real
        cv::Mat img_invert_fourier;
        cv::dft(img_fourier, img_invert_fourier, cv::DFT_REAL_OUTPUT + cv::DFT_INVERSE + cv::DFT_SCALE);

        // convert float dataType to int dataType
        cv::Mat image_int;
        img_invert_fourier.convertTo(image_int, CV_8U);
        cv::imshow("FourierImg", image_int);

        // ------------ 快速离散傅里叶变换 ------------
        cv::Mat img_fft;
        FFT2Image(image_float, img_fft);

        cv::Mat img_ifft;
        cv::dft(img_fft, img_ifft, cv::DFT_REAL_OUTPUT + cv::DFT_INVERSE + cv::DFT_SCALE);
        // 通过裁剪傅里叶逆变换的实部获取的结果 等同原来的输入图像
        cv::Mat img = img_ifft(cv::Rect(0, 0, image.cols, image.rows)).clone();

        img_ifft.convertTo(img_ifft, CV_8U);
        cv::imshow("FFT_Img", img_ifft);

        // ------------ 傅里叶变换的两个度量: 幅度谱和相位谱 ------------
        cv::Mat amplitude_spectrum;
        AmplitudeSpectrumFFT(img_fft, amplitude_spectrum);
        amplitude_spectrum = graySpectrum(amplitude_spectrum);
        cv::imshow("AmplitudeSpectrum", amplitude_spectrum);

        cv::Mat phase_spectrum = PhaseSpectrum(img_fft);
        cv::imshow("PhaseSpectrum", phase_spectrum);

        cv::waitKey(0);
        cv::destroyAllWindows();
        return 0;
    }
    else
    {
        std::cout << "Usage: OpenCV python script imageFile." << std::endl;
        return -1;
    }
}

File: DigitalImageProcessing/python/fourier_transform.py
#!/usr/bin/env python3
# encoding: utf-8

"""
@Funciton: 二维离散傅里叶变换; 快速傅里叶变换; 幅度谱(零谱中心化)和相位谱
@Python Version: 3.8
@Author: Wei Li
@Date: 2021-10-21
"""

import sys
import math
import numpy as np
import cv2 as cv


def FFT2Image(src):
    rows, cols = src.shape[:2]
    # 获取快速傅里叶变换的最优扩充
    row_padding = cv.getOptimalDFTSize(rows)
    col_padding = cv.getOptimalDFTSize(cols)

    # 下侧面和右侧面进行 zero-padding
    img_fft = np.zeros((row_padding, col_padding, 2), np.float32)
    img_fft[:rows, :cols, 0] = src

    # 快速傅里叶变换
    cv.dft(img_fft, img_fft, cv.DFT_COMPLEX_OUTPUT)
    return img_fft


# ------------ 傅里叶变换中两个度量: 幅度谱和相位谱 ------------
def AmplitudeSepectrum(img_fft):
    real_part = np.power(img_fft[:, :, 0], 2.0)
    imaginary_part = np.power(img_fft[:, :, 1], 2.0)
    amplitude_part = np.sqrt(real_part + imaginary_part)

    return amplitude_part


# 将计算出的幅度规格化为灰度级显示，幅度矩阵中的值大部分比较大，往往大于 255,
# 如果只是简单地截断为 255 显示，那么幅度谱呈现的信息会很少。
# 一般采用对数函数对幅度谱进行数值压缩，再进行归一化，
# 这样得到的幅度谱的灰度级显示的对比度会比较高。
def graySpectrum(amplitude):
    # 对比度拉伸
    amplitude_log = np.log(amplitude + 1.0)
    # 归一化
    spectrum_norm = np.zeros(amplitude_log.shape, np.float32)
    cv.normalize(amplitude_log, spectrum_norm, 0, 1, cv.NORM_MINMAX)

    return spectrum_norm


# 对于相位谱的计算，利用 Numpy 中的函数 arctan2，
# 该函数的第一个参数是输入的虚部矩阵，
# 第二个参数是输人的实部矩阵，
# 返回值为对应位置的相位角，数值范围为[-pi, pi]，
# 可以将返回的矩阵除以 pi 再乘以 180，规格化到 [-180, 180]
def phaseSpectrum(fft_img):
    rows, cols = fft_img.shape[:2]
    # 计算对应的相位角
    phase_angle = np.arctan2(fft_img[:, :, 1], fft_img[:, :, 0])
    # 将相位角进行转换
    phase_spectrum = phase_angle / math.pi * 180

    return phase_spectrum



# --------------------------
if __name__ == "__main__":
    if len(sys.argv) > 1:
        image = cv.imread(sys.argv[1], 0)
        if image is None:
            print(f"Error: no such file or dictory.")
        cv.imshow("OriginImage", image)

        # ------------ 傅里叶变换 ------------
        img_fft = FFT2Image(image)

        # invert fast Fourier Transform
        img_ifft = np.zeros(img_fft.shape[:2], np.float32)
        cv.dft(img_fft, img_ifft, cv.DFT_INVERSE + cv.DFT_REAL_OUTPUT + cv.DFT_SCALE)

        # 裁剪
        image_ifft = np.copy(img_ifft[:image.shape[0], :image.shape[1]])
        #裁剪后的结果 image_ifft 等于 image, 傅里叶变换是无损操作
        #通过判断原矩阵减去逆变换裁剪后的矩阵是否为零矩阵，验证两者是否相同
        print(f"The operator of Fourier Transform is ?----{np.max(image_ifft - image)}")
        cv.imshow("FFT_Img", image_ifft.astype(np.uint8))

        # ------------ 傅里叶变换中两个度量: 幅度谱和相位谱 ------------
        amplitude_spectrum = AmplitudeSepectrum(img_fft)
        # 幅度谱的灰度级显示
        ampSpectrum = graySpectrum(amplitude_spectrum)
        cv.imshow("AmplitudeSpectrumGrayLevel", ampSpectrum)
        # 相位谱的灰度级显示
        phase_spectrum = phaseSpectrum(img_fft)
        cv.imshow("PhaseSpectrumGrayLevel", phase_spectrum)

        # -------- 傅里叶变换的幅度谱的中心化 零谱的位置(左上角还是中心)--------
        # 仔细观察会发现一个有趣的现象 —— 中心化后的傅里叶谱比较亮的区域大致与原图中的主要目标垂直
        # stpe 1, 矩阵乘以 (-1)^(rows + cols)
        rows, cols = image.shape
        centralization_amplitude_spectrum = np.copy(image)
        centralization_amplitude_spectrum = centralization_amplitude_spectrum.astype(np.float32)
        for row in range(rows):
            for col in range(cols):
                if (row + col) % 2:
                    centralization_amplitude_spectrum[row][col] = -1*image[row][col]
        # step 2, 快速傅里叶变换
        imgFFT = FFT2Image(image)
        # step 3, 计算幅度谱
        amSpec = AmplitudeSepectrum(imgFFT)
        gray_amSpec = graySpectrum(amSpec)
        cv.imshow("GrayAmSpec", gray_amSpec)

        gray_amspec = 255 * gray_amSpec
        cv.imshow("grayImgSpec", gray_amspec.astype(np.uint8))

        cv.waitKey()
        cv.destroyAllWindows()

    else:
        print("Usage: python python-scripy.py imageFile")
